lib/autoyast.pm

```diff
@@ -207,125 +207,287 @@ sub init_autoyast_profile {
+                if (my $t = get_traversable($node->{$k})) {
+                    # prepend all the expression that will generated for nested nodes
+                    # (in this case with starting point as the 'traversable' node)
+                    # with available info for current child in this iteration.
+                    push @inner_expressions,
+                      map { '/' . ns($t) . $_ } generate_expressions($node->{$k}->{$t});
+                }
+            } else {
+                # continue traversing and accumulating
+                push @inner_expressions, generate_expressions($node->{$k});
             }
+            # all the accumulated inner expressions are concatenated with current node info
+            push @expressions, map { '/' . ns($k) . $_ } @inner_expressions;
         }
     }
+    # the other choice is when is a array ref
     elsif (ref $node eq 'ARRAY') {
-        my ($list_name) = keys %{$node->[0]};
-        my $list_size = scalar @{$node};
-
-        # add expression to check expected list size
-        push @{$expressions}, "$exp" . "[count(ns:$list_name)=$list_size]";
-
         for my $item (@{$node}) {
-            if (my $unique_exp = get_unique_exp(list_name => $list_name, item => $item,
-                    exp => $exp, errors => $errors)) {
-                push @{$expressions}, $unique_exp;
-                generate_expressions(node => $item->{$list_name}, exp => $unique_exp,
-                    expressions => $expressions, errors => $errors);
+            # only in case we have a list of scalars
+            if (!ref $item) {
+                push @expressions, create_xpath_predicate($item);
+                next;
+            }
+            # get items with some nested properties
+            my @processables = grep { is_processable($item->{$_}) } sort keys(%{$item});
+            # create a predicate and consider where to look, if direct children or any descendant
+            my @predicates = map {
+                get_descendant($item->{$_}) . ns($_) . create_xpath_predicate($item->{$_})
+            } @processables;
+            # close the predicate joining all the intermediate ones and add it as expression
+            my $predicate = close_predicate(@predicates);
+            push @expressions, $predicate;
+
+            # consider how to continue traversing the YAML when we need to search a node to traverse
+            # in case it could have some properties
+            for my $p (@processables) {
+                if (my $t = get_traversable($item->{$p})) {
+                    # concatenate current node info with result of recursive call
+                    push @expressions, map { $predicate . '/' . ns($p) . '/' . ns($t) . $_ } generate_expressions($item->{$p}->{$t});
+                }
+            }
+
+            # items which do not have properties are directly ready to traverse them
+            my @traversables = grep { !is_processable($item->{$_}) } sort keys(%{$item});
```

Comment:
Same here, I don't think that we need to spend cpu time on sorting keys.

Reply:
This is for visualization, when checking the output I saw that improved readability of the XPATH expressions.

Reply:
Ah, I see. We also could sort them once before printing, as we do so only in case of failure.

Reply:
the order of the keys in the predicates helps to debug in the non-failing cases as well. To change the order once the whole expression has been generated would be very tricky, my plan is beatify the string with the keys in the order that they appear.